File: projects/py/src/sonic_ws/schema_validation.py
from .packet_type import PacketType

UNSIGNED = {PacketType.UBYTES, PacketType.USHORTS, PacketType.UVARINT}
NUMERIC = {
    PacketType.BYTES, PacketType.UBYTES, PacketType.SHORTS, PacketType.USHORTS,
    PacketType.VARINT, PacketType.UVARINT, PacketType.DELTAS,
    PacketType.FLOATS, PacketType.DOUBLES,
}


def validate_packet_schema(packets, *, direction=None, warn_unbounded=False):
    packets = list(packets)
    errors, warnings = [], []
    tags, variants = set(), set()
    parents = {packet.tag: packet for packet in packets if packet.is_parent}
    if len(packets) > 254:
        errors.append(f"packet table contains {len(packets)} packets; the maximum is 254")

    for packet in packets:
        if packet.tag in tags:
            errors.append(f'duplicate packet tag "{packet.tag}"')
        tags.add(packet.tag)
        if packet.replay and packet.data_batching:
            errors.append(f'packet "{packet.tag}" combines replay with batching')
        if packet.quantized and any(kind not in NUMERIC for kind in packet.types):
            errors.append(f'packet "{packet.tag}" quantizes a non-numeric type')
        if packet.value_min is not None and packet.value_min < 0 and any(kind in UNSIGNED for kind in packet.types):
            errors.append(f'packet "{packet.tag}" has a negative minimum for an unsigned type')
        if warn_unbounded and direction == "client" and any(value >= 2_048_383 for value in packet.data_maxes):
            warnings.append(f'client packet "{packet.tag}" has an effectively unbounded value count')
        if packet.parent and packet.variant:
            key = f"{packet.parent}.{packet.variant}"
            if key in variants:
                errors.append(f'duplicate packet-group variant "{key}"')
            variants.add(key)

    for packet in packets:
        if not packet.parent:
            continue
        parent = parents.get(packet.parent)
        if parent is None:
            errors.append(f'packet "{packet.tag}" references missing group parent "{packet.parent}"')
        elif parent.type != PacketType.NONE:
            errors.append(f'packet-group parent "{packet.parent}" must use PacketType.NONE')
    return {"errors": errors, "warnings": warnings}


def assert_packet_schema(packets, **options):
    result = validate_packet_schema(packets, **options)
    if result["errors"]:
        raise ValueError("invalid SonicWS packet schema:\n- " + "\n- ".join(result["errors"]))
    return result


ValidatePacketSchema = validate_packet_schema
AssertPacketSchema = assert_packet_schema

File: projects/ts/src/ws/util/packets/metadata/SchemaValidation.ts
import { Packet } from "../../../packets/Packets";
import { PacketType } from "../../../packets/PacketType";

export type SchemaValidationResult = { errors: string[]; warnings: string[] };

const UNSIGNED = new Set([PacketType.UBYTES, PacketType.USHORTS, PacketType.UVARINT]);
const NUMERIC = new Set([
    PacketType.BYTES, PacketType.UBYTES, PacketType.SHORTS, PacketType.USHORTS,
    PacketType.VARINT, PacketType.UVARINT, PacketType.DELTAS, PacketType.FLOATS, PacketType.DOUBLES,
]);

/** Performs whole-table checks that individual CreatePacket calls cannot see. */
export function ValidatePacketSchema(
    packets: readonly Packet<any>[],
    options: { direction?: "client" | "server"; warnUnbounded?: boolean } = {},
): SchemaValidationResult {
    const errors: string[] = [];
    const warnings: string[] = [];
    const tags = new Set<string>();
    const parents = new Map<string, Packet<any>>();
    const variants = new Set<string>();

    if (packets.length > 254) {
        errors.push(`Packet table contains ${packets.length} packets; the maximum is 254`);
    }

    for (const packet of packets) {
        if (tags.has(packet.tag)) {
            errors.push(`Duplicate packet tag "${packet.tag}"`);
        }
        tags.add(packet.tag);

        if (packet.isParent) {
            parents.set(packet.tag, packet);
        }
        if (packet.replay && packet.dataBatching) {
            errors.push(`Packet "${packet.tag}" combines replay with batching`);
        }

        const types = Array.isArray(packet.type) ? packet.type : [packet.type];
        if (packet.quantized && types.some(type => !NUMERIC.has(type))) {
            errors.push(`Packet "${packet.tag}" quantizes a non-numeric type`);
        }
        if (packet.valueMin !== undefined && packet.valueMin < 0 && types.some(type => UNSIGNED.has(type))) {
            errors.push(`Packet "${packet.tag}" has a negative minimum for an unsigned type`);
        }
        if (
            packet.fields
            && !packet.autoFlatten
            && !packet.object
            && packet.dataMin === packet.dataMax
            && packet.fields.length !== packet.dataMax
        ) {
            errors.push(`Packet "${packet.tag}" schema length does not match its fixed value count`);
        }
        const unbounded = Array.isArray(packet.dataMax)
            ? packet.dataMax.some(value => value >= 2_048_383)
            : packet.dataMax >= 2_048_383;
        if (options.warnUnbounded && options.direction === "client" && unbounded) {
            warnings.push(`Client packet "${packet.tag}" has an effectively unbounded value count`);
        }

        if (packet.parent && packet.variant) {
            const key = `${packet.parent}.${packet.variant}`;
            if (variants.has(key)) {
                errors.push(`Duplicate packet-group variant "${key}"`);
            }
            variants.add(key);
        }
    }

    for (const packet of packets) {
        if (!packet.parent) continue;
        const parent = parents.get(packet.parent);
        if (!parent) {
            errors.push(`Packet "${packet.tag}" references missing group parent "${packet.parent}"`);
        } else if (parent.type !== PacketType.NONE) {
            errors.push(`Packet-group parent "${packet.parent}" must use PacketType.NONE`);
        }
    }

    return { errors, warnings };
}

/** Throws when a packet table contains structural schema errors. */
export function AssertPacketSchema(
    packets: readonly Packet<any>[],
    options?: Parameters<typeof ValidatePacketSchema>[1],
): void {
    const result = ValidatePacketSchema(packets, options);
    if (result.errors.length) {
        throw new Error(`Invalid SonicWS packet schema:\n- ${result.errors.join("\n- ")}`);
    }

    for (const warning of result.warnings) {
        console.warn(`SonicWS schema warning: ${warning}`);
    }
}
